### PR TITLE
TST: MultiIndex.get_indexer with na/missing

### LINKED ISSUE #29252: Bug: Joining MultiIndex with NaNs
#### Code Sample, a copy-pastable example if possible

```python
import numpy as np
import pandas as pd

df1 = pd.DataFrame(data={'col1': [1.1, 1.2]},
                   index=pd.MultiIndex.from_product([['A'], [1.0, 2.0]],
                                                    names=['id1', 'id2']))
df2 = pd.DataFrame(data={'col2': [2.1, 2.2]},
                   index=pd.MultiIndex.from_product([['A'], [np.NaN, 2.0]],
                                                    names=['id1', 'id2']))
print(df1.join(df2))
```
#### Problem description

The index containing the NaN value of `df2` is falsely joined with the index of `df1`. Therefore, the result contains both values of `df2` instead of only one value and a NaN value.

#### Expected Output

```
         col1  col2
id1 id2
A   1.0   1.1   NaN
    2.0   1.2   2.2
```

The expected result was returned by older pandas versions (successfully tested with 0.18.1 ; 0.20.3 ; 0.21.1 ; 0.22.0).

#### Actual Output

```
         col1  col2
id1 id2
A   1.0   1.1   2.1
    2.0   1.2   2.2
```

This result is returned by newer pandas versions, beginning with version 0.23.0 (tested with 0.23.0 ; 0.23.4 ; 0.24.2 ; 0.25.2).

I suppose this is a bug, or am I missing something?

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.2.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 94 Stepping 3, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.None

pandas           : 0.25.2
numpy            : 1.16.2
pytz             : 2018.9
dateutil         : 2.8.0
pip              : 19.0.3
setuptools       : 40.8.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
</details>

### LINKED ISSUE #37222: BUG: get_indexer for MultiIndex with nans returns wrong indexer
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
idx1 = pd.MultiIndex.from_product([['A'], [1.0, 2.0]], names=['id1', 'id2'])
idx2 = pd.MultiIndex.from_product([['A'], [np.nan, 2.0]], names=['id1', 'id2'])


print(idx2.get_indexer(idx1))
print(idx1.get_indexer(idx2))

```

#### Problem description

This snippet returns 
```
[0 1]
[-1  1]
```

#### Expected Output

It should return ``[-1, 1]`` for both statement.

Part of the problem is in 
https://github.com/pandas-dev/pandas/blob/54fa3da2e0b65b733d12de6cbecbf2d8afef4e9f/pandas/core/indexes/multi.py#L1045-L1064

The ``self.levels`` statement does not hand the ``nans`` over. But even fixing this returns 
```
[ 0 -1]
[-1  1]
```

The error must be somewhere deeper, but I could not figure out where something goes wrong. Tested on 1.0.5 and 0.25.3. Does not seem to be a regression.

#### Output of ``pd.show_versions()``

<details>

master

</details>

### LINKED ISSUE #38623: BUG: Index and MultiIndex ops treating nan differently
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

``Index.union`` treats ``nan`` differently than ``MultiIndex.union``. Same goes for intersection. 

```python
idx1 = Index([1, np.nan])
idx2 = Index([1, np.nan, 3])

idx1.union(idx2)
idx1.intersection(idx2)

mi1 = MultiIndex.from_arrays([[1, np.nan]])
mi2 = MultiIndex.from_arrays([[1, np.nan, 3]])
mi1.union(mi2)
mi1.intersection(mi2)

```

Output:
```
Float64Index([1.0, 3.0, nan], dtype='float64')
Float64Index([1.0, nan], dtype='float64')
MultiIndex([(1.0,),
            (nan,),
            (nan,),
            (3.0,)],
           )
MultiIndex([(1.0,)],
           )
```

#### Problem description

Inconsistent behavior is a problem. Additionally if both MultiIndexes are *equal* nan is only returned once. When not matching nan is the desired output, we have to adjust for theses cases before checking for equality. 

#### Expected Output

I am not quite sure what to expect, but taking the current behavior of merges into account, I think it would be more reasonable to return
```
MultiIndex([(1.0,),
            (nan,),
            (3.0,)],
           )
MultiIndex([(1.0,),
            (nan,)],
           )
```

for the MultiIndex operations. Thoughts @jbrockmendel ?

related to #38439

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : c6b05c3ad771da775011647f66687a398e11c533
python           : 3.8.6.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.0-58-generic
Version          : #64-Ubuntu SMP Wed Dec 9 08:16:25 UTC 2020
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.0.dev0+129.gc6b05c3ad7
numpy            : 1.19.2
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.2.3
setuptools       : 49.6.0.post20201009
Cython           : 0.29.21
pytest           : 6.1.1
hypothesis       : 5.37.1
sphinx           : 3.2.1
blosc            : None
feather          : None
xlsxwriter       : 1.3.7
lxml.etree       : 4.5.2
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.18.1
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : 1.3.2
fsspec           : 0.8.3
fastparquet      : 0.4.1
gcsfs            : 0.7.1
matplotlib       : 3.3.2
numexpr          : 2.7.1
odfpy            : None
openpyxl         : 3.0.5
pandas_gbq       : None
pyarrow          : 2.0.0
pyxlsb           : None
s3fs             : 0.4.2
scipy            : 1.5.2
sqlalchemy       : 1.3.20
tables           : 3.6.1
tabulate         : 0.8.7
xarray           : 0.16.1
xlrd             : 1.2.0
xlwt             : 1.3.0
numba            : 0.51.2
None

Process finished with exit code 0

</details>

### LINKED ISSUE #42883: BUG: Using None in nested object MultiIndex causes reindex to misalign.
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
>>> import pandas as pd
>>> index = pd.MultiIndex.from_tuples([(('a', None), 1), (('b', None), 2)])
>>> index2 = pd.MultiIndex.from_tuples([(('b', None), 2), (('a', None), 1)])
>>> df1_dtype = pd.DataFrame([1,2], index=index)
>>> df2_dtype = pd.DataFrame([2,1], index=index2)
>>> df1_dtype.reindex_like(df2_dtype)
              0
(b, None) 2 NaN
(a, None) 1 NaN
```

```python
>>> import pandas as pd
>>> import numpy as np
>>> index = pd.MultiIndex.from_tuples([(('a', np.nan), 1), (('b', np.nan), 2)])
>>> index2 = pd.MultiIndex.from_tuples([(('b', np.nan), 2), (('a', np.nan), 1)])
>>> df1_dtype = pd.DataFrame([1,2], index=index)
>>> df2_dtype = pd.DataFrame([2,1], index=index2)
>>> df1_dtype.reindex_like(df2_dtype)
            0
(b, nan) 2  2
(a, nan) 1  1
```

#### Problem description

I would expect the two indices to align here.

#### Expected Output
```python
>>> import pandas as pd
>>> index = pd.MultiIndex.from_tuples([(('a', None), 1), (('b', None), 2)])
>>> index2 = pd.MultiIndex.from_tuples([(('b', None), 2), (('a', None), 1)])
>>> df1_dtype = pd.DataFrame([1,2], index=index)
>>> df2_dtype = pd.DataFrame([2,1], index=index2)
>>> df1_dtype.reindex_like(df2_dtype)
            0
(b, None) 2  2
(a, None) 1  1
```
#### Output of ``pd.show_versions()``

<details>

```python3
>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit           : 226876a55b82324b92449256bab47d431b22310f
python           : 3.8.5.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.19.128-microsoft-standard
Version          : #1 SMP Tue Jun 23 12:58:10 UTC 2020
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : C.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.4.0.dev0+357.g226876a55b
numpy            : 1.21.1
pytz             : 2021.1
dateutil         : 2.8.2
pip              : 21.2.2
setuptools       : 57.4.0
Cython           : 0.29.24
pytest           : 6.2.4
hypothesis       : 6.14.5
sphinx           : 4.1.2
blosc            : 1.10.4
feather          : None
xlsxwriter       : 1.4.4
lxml.etree       : 4.6.3
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 3.0.1
IPython          : 7.25.0
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : 1.3.2
fsspec           : 2021.05.0
fastparquet      : 0.6.3
gcsfs            : 2021.07.0
matplotlib       : 3.4.2
numexpr          : 2.7.3
odfpy            : None
openpyxl         : 3.0.7
pandas_gbq       : None
pyarrow          : 5.0.0
pyxlsb           : None
s3fs             : None
scipy            : 1.7.0
sqlalchemy       : 1.4.22
tables           : 3.6.1
tabulate         : 0.8.9
xarray           : 0.18.2
xlrd             : 2.0.1
xlwt             : 1.3.0
numba            : 0.53.1
```

</details>

### LINKED ISSUE #43222: BUG: Constructing DataFrame from list of series with MultiIndex, if MultiIndex contains (NaN, NaN)
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

I have observed a few issues, when constructing a dataframe from a list of series with multiindices, if any multiindex contains (NaN, NaN).

#### Code Sample, a copy-pastable example

1. In this case I would expect the columns of the dataframe to be the union of the indices of series1 and series2, but ("a", "b") is missing.
```python
import pandas as pd
series1 = pd.Series((1,), index=pd.MultiIndex.from_tuples(((None, None),)))
series2 = pd.Series((10, 20), index=pd.MultiIndex.from_tuples(((None, None), ("a", "b"))))
df = pd.DataFrame([series1, series2])
#   NaN
#   NaN
# 0   1
# 1  10
```

2. There is another issue, when series1 and series2 are flipped.  The columns appear to be correct, but in this case I would expect the value NaN at index 1, column (a,b)):
```python
df = pd.DataFrame([series2, series1])
#     a NaN
#     b NaN
# 0  20  10
# 1   1   1
```

#### Expected Output
```python
df = pd.DataFrame([series1, series2])
#     a NaN
#     b NaN
# 0  NaN 1
# 1  20  10
```

and the reverse case

```python
df = pd.DataFrame([series2, series1])
#     a NaN
#     b NaN
# 0  20  10
# 1  NaN 1
```

This behavior is not limited to both series containing the MultiIndex (NaN, NaN).  Here is an example, where only one series is constructed with (None, None).  The column ("a", "b") is missing.

```python
import pandas as pd
series1 = pd.Series((1,), index=pd.MultiIndex.from_tuples((("a", None),)))
series2 = pd.Series((10, 20), index=pd.MultiIndex.from_tuples(((None, None), ("a", "b"))))
df = pd.DataFrame([series1, series2])
#      a   NaN
#    NaN   NaN
# 0  1.0   NaN
# 1  NaN  10.0
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : f346574c8bb8921664f6d47d2e4b0f1252e181b9
python           : 3.9.6.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.10.59-1-MANJARO
Version          : #1 SMP PREEMPT Sun Aug 15 13:11:32 UTC 2021
machine          : x86_64
processor        : 
byteorder        : little
LC_ALL           : None
LANG             : en_US.utf8
LOCALE           : en_US.UTF-8

pandas           : 0+untagged.1.gf346574
numpy            : 1.21.1
pytz             : 2021.1
dateutil         : 2.8.2
pip              : 21.1.3
setuptools       : 57.4.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 7.26.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.4.3
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

</details>

### LINKED ISSUE #46173: Missing keys in a selector list are matched to None-labeled entries of MultiIndex
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
df = pd.DataFrame.from_dict({
    ("foo",): [1, 2, 3],
    ("bar",): [5, 6, 7],
    (None,): [8, 9, 0], 
})
df[[ ("missingKey",) ]] # returns the NaN-labeled column [8, 9, 0] instead of raising a KeyError
```

### Issue Description

Given a DataFrame with MultiIndex containing NaN values in its keys, then keys with missing labels on the same level as NaN values will retrieve the NaN-labeled columns if keys are passed in a list.

If multiple missing labels are passed, each of them will retrieve the None column
```python
pd.DataFrame.from_dict({(None,): [8, 9, 0]}).loc[:, [("foo",), ("bar",)]]  # returns a DF with two copies of the NaN-labeled [8, 9, 0] column 
``` 

This behaviour occurs only when selection is done via a list of keys. 
```python
pd.DataFrame.from_dict({(None,): [8, 9, 0]}).loc[:, ("foo",)]  # single key - raises KeyError
pd.DataFrame.from_dict({(None,): [8, 9, 0]}).loc[:, [("foo",)]]  # key in a list - returns [8,9,0]
```

The same issue occurs for multi-level MultiIndex, as long as missing labels in the selector occur only on the same level as NaN values.
```python
pd.DataFrame.from_dict({(None,None): [8, 9, 0]}).loc[:, [("a","a")]]  # returns [8, 9, 0]
pd.DataFrame.from_dict({(None,None): [8, 9, 0]}).loc[:, [("a","b")]]  # returns [8, 9, 0]
pd.DataFrame.from_dict({(None,"b"): [8, 9, 0]}).loc[:, [("a","b")]]  # returns [8, 9, 0]
pd.DataFrame.from_dict({(None,"a"): [8, 9, 0]}).loc[:, [("a","b")]]  # KeyError, because 2nd level didn't match
```

`None` can also be replaced with any of `pd.NA`, `np.nan`, `pd.NaT` for the same effect. 

An additional, more complex example, with multiple `None` in index.
```python
df = pd.DataFrame.from_dict({
    (None,None): [1, 2, 3],
    (None,"a"): [5, 6, 7],
    ("b",None): [8, 9, 0],
})

df[[("foo", "bar")]] # returns [1,2,3]
df[[("foo", "a")]] # returns [5,6,7]
df[[("b", "foo")]] # returns [8,9,0]
df[[("b", "a")]] # raises KeyError
```
Interestingly enough, if `[("a", "b")]` is used to select from this DF, a KeyError *will* be raised, instead of matching it to (None, None).

### Expected Behavior

A KeyError should be raised, as it would be if the key with missing label was not wrapped in a list.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 06d230151e6f18fdb8139d09abf539867a8cd481
python           : 3.8.10.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.13.0-27-generic
Version          : #29~20.04.1-Ubuntu SMP Fri Jan 14 00:32:30 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 1.4.1
numpy            : 1.22.2
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 20.0.2
setuptools       : 44.0.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.6.3
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.3
IPython          : 8.0.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.4.1
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None

</details>

### LINKED ISSUE #48905: BUG: regression in MultiIndex.isin when index contains NA
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd

mi1 = pd.MultiIndex.from_tuples([(1, pd.NA)])
mi2 = pd.MultiIndex.from_tuples([(1, 1), (1, 2)])

mi2.isin(mi1)
```

### Issue Description

output:
```
array([True, True])    # <- main
array([False, False])  # <- 1.5.0
```

Regression in on main only and looks to be a result of #48622. 

The bug is not in `MultiIndex.isin` per se, but #48622 looks to have exposed the underlying bug to `MultiIndex.isin`.

Open PR #48877 will fix the underlying bug. 

### Expected Behavior

```
array([False, False])
```

### Installed Versions

<details>

Replace this line with the output of pd.show_versions()

</details>

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] closes #43222
- [x] closes #46173
- [x] closes #42883
- [x] closes #38623
- [x] closes #29252
- [x] closes #48905 
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v2.0.0.rst` file if fixing a bug or adding a new feature.

Update: #49442 has been merged to fix the underlying bug here. This PR has been updated to add tests for closing the linked issues which are all related to the same underlying bug in #49442.

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -4028,6 +4028,11 @@ def get_indexer(
                 target, method=method, limit=limit, tolerance=tolerance
             )
 
+        if self._is_multi and self._contains_na_any_level:
```
Comment (phofl) [inline pandas/core/indexes/base.py:4031]: Id prefer fixing this on the engine level and not special casing

Comment (lukemanley) [reply pandas/core/indexes/base.py:4031]: ok, i'll take another look

Comment (phofl) [reply pandas/core/indexes/base.py:4031]: This is probably non trivial, not sure what's wrong there

Comment (lukemanley) [reply pandas/core/indexes/base.py]: not seeing an obvious way to fix this on the engine level. If you look a few lines above this, categorical is already special cased to handle nans in a similar way. Are you ok with special casing for now?

**Comment by phofl** [comment]
Could you rebase? Hopefully, all tests should pass now. Also, would be good to add the issues to my whatsnew note. wdyt?

**Comment by lukemanley** [comment]
> Could you rebase? Hopefully, all tests should pass now. Also, would be good to add the issues to my whatsnew note. wdyt?

Updated this PR to add tests only and close linked issues. If you think any of these tests are redundant, I'm happy to remove some (all reported for the same underlying bug).

**Review by mroeschke — APPROVED** [review]
LGTM. Merge when ready @phofl

**Review by phofl — APPROVED** [review]
Thx, I like the tests, since there is now guarantee that the underlying implementation won’t change in the future. So covering our bases should be good

**Pull request merged**